File: backend/app/routes/weather.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession
import httpx
from datetime import datetime, timedelta
from typing import Optional, Dict, Any
from ..config import settings
from ..database import get_db

router = APIRouter()

def get_fallback_weather(city: str) -> Dict[str, Any]:
    # Dynamic deterministic weather based on city name hash code
    city_hash = sum(ord(c) for c in city)
    
    # Calculate condition
    conditions = ["साफ़ (Sunny)", "बादल (Cloudy)", "हल्की बारिश (Light Rain)", "तूफान (Thunderstorm)", "धुंध (Mist)"]
    cond_index = city_hash % len(conditions)
    condition = conditions[cond_index]
    
    # Calculate temp
    temp = 25 + (city_hash % 15) # 25°C to 39°C
    humidity = 40 + (city_hash % 50) # 40% to 90%
    wind_speed = 5 + (city_hash % 20) # 5 to 25 km/h
    pressure = 1005 + (city_hash % 10)
    visibility = 6 + (city_hash % 5)
    
    # Generate 7 day forecast
    forecast = []
    days_hi = ["रविवार", "सोमवार", "मंगलवार", "बुधवार", "गुरुवार", "शुक्रवार", "शनिवार"]
    current_day = datetime.now()
    
    for i in range(7):
        forecast_day = current_day + timedelta(days=i)
        day_name = "आज" if i == 0 else "कल" if i == 1 else "परसों" if i == 2 else days_hi[forecast_day.weekday()]
        
        day_hash = city_hash + i
        day_cond = conditions[day_hash % len(conditions)]
        high = temp + (day_hash % 4) - 2
        low = high - 8 - (day_hash % 3)
        rain_chance = 0 if "sunny" in day_cond.lower() else 20 + (day_hash % 80)
        
        forecast.append({
            "day": day_name,
            "condition": day_cond,
            "high": int(high),
            "low": int(low),
            "rain_chance": rain_chance
        })
        
    # Generate farming tips
    tips = []
    if "rain" in condition.lower() or "thunderstorm" in condition.lower():
        tips.append({"type": "danger", "text": "🌧️ भारी बारिश की संभावना है — अपनी फसल कटाई रोकें और भीगने से बचाएं।"})
        tips.append({"type": "warning", "text": "💧 खेतों में अतिरिक्त पानी निकासी का प्रबंध करें, सिंचाई तुरंत बंद करें।"})
    elif temp > 35:
        tips.append({"type": "warning", "text": "🥵 अत्यधिक तापमान — फसलों को झुलसने से बचाने के लिए शाम को हल्की सिंचाई करें।"})
        tips.append({"type": "safe", "text": "🌻 गर्मी सहन करने वाली फसलें (जैसे मक्का) के लिए अनुकूल मौसम।"})
    else:
        tips.append({"type": "safe", "text": "🌤️ मौसम सुहावना है — आज सिंचाई और उर्वरक छिड़काव (Urea spray) के लिए उत्तम दिन है।"})
        tips.append({"type": "safe", "text": "🌾 कटी हुई फसल को धूप में सुखाने का सही समय है।"})
        
    return {
        "location": city,
        "date": datetime.now().strftime("%d %B %Y"),
        "temp": int(temp),
        "condition": condition,
        "humidity": humidity,
        "wind_speed": wind_speed,
        "pressure": pressure,
        "visibility": visibility,
        "forecast": forecast,
        "farming_tips": tips
    }

@router.post("/")
async def get_weather(location: str = Query(...)):
    location = location.strip()
    if not location:
        raise HTTPException(status_code=400, detail="Location is required")
        
    if not settings.OPENWEATHER_API_KEY:
        # Return fallback mock data
        return {"weather": get_fallback_weather(location)}
        
    # If API key is present, call OpenWeather
    try:
        async with httpx.AsyncClient() as client:
            # First fetch lat/lon
            geo_url = f"http://api.openweathermap.org/geo/1.0/direct?q={location},IN&limit=1&appid={settings.OPENWEATHER_API_KEY}"
            geo_res = await client.get(geo_url)
            geo_data = geo_res.json()
            
            if not geo_data:
                return {"weather": get_fallback_weather(location)}
                
            lat, lon = geo_data[0]["lat"], geo_data[0]["lon"]
            city_name = geo_data[0]["name"]
            
            # Fetch weather data
            weather_url = f"https://api.openweathermap.org/data/2.5/forecast?lat={lat}&lon={lon}&units=metric&appid={settings.OPENWEATHER_API_KEY}"
            weather_res = await client.get(weather_url)
            weather_data = weather_res.json()
            
            # Parse weather data
            list_data = weather_data["list"]
            current = list_data[0]
            
            condition_map = {
                "clear": "साफ़ (Sunny)",
                "clouds": "बादल (Cloudy)",
                "rain": "बारिश (Rain)",
                "drizzle": "हल्की बारिश (Drizzle)",
                "thunderstorm": "तूफान (Thunderstorm)",
                "snow": "बर्फ़बारी",
                "mist": "धुंध (Mist)",
                "smoke": "धुआं",
                "haze": "धुंध (Haze)",
                "dust": "धूल",
                "fog": "कोहरा"
            }
            raw_cond = current["weather"][0]["main"].lower()
            condition = condition_map.get(raw_cond, current["weather"][0]["description"])
            
            # Generate 7 days forecast from list (every 8th item is ~24 hours later)
            forecast = []
            days_hi = ["रविवार", "सोमवार", "मंगलवार", "बुधवार", "गुरुवार", "शुक्रवार", "शनिवार"]
            
            for idx in range(0, min(len(list_data), 56), 8):
                day_data = list_data[idx]
                day_time = datetime.fromtimestamp(day_data["dt"])
                day_name = "आज" if idx == 0 else "कल" if idx == 8 else "परसों" if idx == 16 else days_hi[day_time.weekday()]
                
                day_raw_cond = day_data["weather"][0]["main"].lower()
                day_cond = condition_map.get(day_raw_cond, day_data["weather"][0]["description"])
                
                forecast.append({
                    "day": day_name,
                    "condition": day_cond,
                    "high": int(day_data["main"]["temp_max"]),
                    "low": int(day_data["main"]["temp_min"]),
                    "rain_chance": int(day_data.get("pop", 0) * 100) # probability of precipitation
                })
                
            # Create farming tips
            tips = []
            rain_upcoming = any("rain" in f["condition"].lower() or "drizzle" in f["condition"].lower() for f in forecast[:3])
            
            if rain_upcoming:
                tips.append({"type": "danger", "text": "🌧️ अगले कुछ दिनों में बारिश की संभावना है — अपनी फसल कटाई रोकें और सुरक्षित स्थान पर रखें।"})
                tips.append({"type": "warning", "text": "💧 सिंचाई तुरंत बंद करें। खेतों में पानी जमा न होने दें।"})
            else:
                temp_max = current["main"]["temp"]
                if temp_max > 35:
                    tips.append({"type": "warning", "text": "🥵 अत्यधिक गर्मी — फसलों को सूखने से बचाने के लिए शाम को हल्की सिंचाई करें।"})
                tips.append({"type": "safe", "text": "🌤️ मौसम अनुकूल है — आज कीटनाशक छिड़काव (Pest spray) या यूरिया का प्रयोग करें।"})
                tips.append({"type": "safe", "text": "🌾 कटी फसलों को सुखाने और मंडी ले जाने के लिए उपयुक्त समय है।"})
                
            return {
                "weather": {
                    "location": city_name,
                    "date": datetime.now().strftime("%d %B %Y"),
                    "temp": int(current["main"]["temp"]),
                    "condition": condition,
                    "humidity": current["main"]["humidity"],
                    "wind_speed": int(current["wind"]["speed"] * 3.6), # m/s to km/h
                    "pressure": current["main"]["pressure"],
                    "visibility": int(current.get("visibility", 10000) / 1000), # m to km
                    "forecast": forecast,
                    "farming_tips": tips
                }
            }
    except Exception as e:
        # Fallback if request fails
        return {"weather": get_fallback_weather(location)}

@router.get("/auto")
async def get_weather_auto(lat: float, lon: float):
    # Reverse geocoding or direct weather lookup
    if not settings.OPENWEATHER_API_KEY:
        return {"city": "वाराणसी"} # Default testing city
        
    try:
        async with httpx.AsyncClient() as client:
            url = f"http://api.openweathermap.org/geo/1.0/reverse?lat={lat}&lon={lon}&limit=1&appid={settings.OPENWEATHER_API_KEY}"
            res = await client.get(url)
            data = res.json()
            if data:
                return {"city": data[0]["name"]}
    except Exception:
        pass
    return {"city": "वाराणसी"}

File: backend/src/routes/weather.js
const express = require('express');
const router = express.Router();
const axios = require('axios');

const getFallbackWeather = (city) => {
  const cityHash = (city || '').split('').reduce((acc, c) => acc + c.charCodeAt(0), 0);
  const conditions = ["साफ़ (Sunny)", "बादल (Cloudy)", "हल्की बारिश (Light Rain)", "तूफान (Thunderstorm)", "धुंध (Mist)"];
  const condIndex = cityHash % conditions.length;
  const condition = conditions[condIndex];

  const temp = 25 + (cityHash % 15);
  const humidity = 40 + (cityHash % 50);
  const windSpeed = 5 + (cityHash % 20);
  const pressure = 1005 + (cityHash % 10);
  const visibility = 6 + (cityHash % 5);

  const daysHi = ["रविवार", "सोमवार", "मंगलवार", "बुधवार", "गुरुवार", "शुक्रवार", "शनिवार"];
  const forecast = [];
  const now = new Date();

  for (let i = 0; i < 7; i++) {
    const forecastDay = new Date(now.getTime() + i * 24 * 60 * 60 * 1000);
    const dayName = i === 0 ? "आज" : i === 1 ? "कल" : i === 2 ? "परसों" : daysHi[forecastDay.getDay()];

    const dayHash = cityHash + i;
    const dayCond = conditions[dayHash % conditions.length];
    const high = temp + (dayHash % 4) - 2;
    const low = high - 8 - (dayHash % 3);
    const rainChance = dayCond.toLowerCase().includes('sunny') ? 0 : 20 + (dayHash % 80);

    forecast.push({
      day: dayName,
      condition: dayCond,
      high: Math.round(high),
      low: Math.round(low),
      rain_chance: rainChance
    });
  }

  const tips = [];
  if (condition.toLowerCase().includes('rain') || condition.toLowerCase().includes('thunderstorm')) {
    tips.push({ type: "danger", text: "🌧️ भारी बारिश की संभावना है — अपनी फसल कटाई रोकें और भीगने से बचाएं।" });
    tips.push({ type: "warning", text: "💧 खेतों में अतिरिक्त पानी निकासी का प्रबंध करें, सिंचाई तुरंत बंद करें।" });
  } else if (temp > 35) {
    tips.push({ type: "warning", text: "🥵 अत्यधिक तापमान — फसलों को झुलसने से बचाने के लिए शाम को हल्की सिंचाई करें।" });
    tips.push({ type: "safe", text: "🌻 गर्मी सहन करने वाली फसलें (जैसे मक्का) के लिए अनुकूल मौसम।" });
  } else {
    tips.push({ type: "safe", text: "🌤️ मौसम सुहावना है — आज सिंचाई और उर्वरक छिड़काव (Urea spray) के लिए उत्तम दिन है।" });
    tips.push({ type: "safe", text: "🌾 कटी हुई फसल को धूप में सुखाने का सही समय है।" });
  }

  const months = ["January", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December"];
  const formattedDate = `${now.getDate()} ${months[now.getMonth()]} ${now.getFullYear()}`;

  return {
    location: city,
    date: formattedDate,
    temp: Math.round(temp),
    condition,
    humidity,
    wind_speed: windSpeed,
    pressure,
    visibility,
    forecast,
    farming_tips: tips
  };
};

router.post('/', async (req, res) => {
  try {
    const location = (req.query.location || req.body.location || 'वाराणसी').trim();

    if (!process.env.OPENWEATHER_API_KEY) {
      return res.json({ weather: getFallbackWeather(location) });
    }

    const geoUrl = `http://api.openweathermap.org/geo/1.0/direct?q=${encodeURIComponent(location)},IN&limit=1&appid=${process.env.OPENWEATHER_API_KEY}`;
    const geoRes = await axios.get(geoUrl);
    if (!geoRes.data || geoRes.data.length === 0) {
      return res.json({ weather: getFallbackWeather(location) });
    }

    const { lat, lon, name: cityName } = geoRes.data[0];
    const weatherUrl = `https://api.openweathermap.org/data/2.5/forecast?lat=${lat}&lon=${lon}&units=metric&appid=${process.env.OPENWEATHER_API_KEY}`;
    const weatherRes = await axios.get(weatherUrl);

    const listData = weatherRes.data.list;
    const current = listData[0];

    const conditionMap = {
      clear: "साफ़ (Sunny)",
      clouds: "बादल (Cloudy)",
      rain: "बारिश (Rain)",
      drizzle: "हल्की बारिश (Drizzle)",
      thunderstorm: "तूफान (Thunderstorm)",
      mist: "धुंध (Mist)",
      fog: "कोहरा"
    };

    const rawCond = (current.weather[0].main || '').toLowerCase();
    const condition = conditionMap[rawCond] || current.weather[0].description;

    const daysHi = ["रविवार", "सोमवार", "मंगलवार", "बुधवार", "गुरुवार", "शुक्रवार", "शनिवार"];
    const forecast = [];

    for (let idx = 0; idx < Math.min(listData.length, 56); idx += 8) {
      const dayData = listData[idx];
      const dayTime = new Date(dayData.dt * 1000);
      const dayName = idx === 0 ? "आज" : idx === 8 ? "कल" : idx === 16 ? "परसों" : daysHi[dayTime.getDay()];
      const dayRawCond = (dayData.weather[0].main || '').toLowerCase();
      const dayCond = conditionMap[dayRawCond] || dayData.weather[0].description;

      forecast.push({
        day: dayName,
        condition: dayCond,
        high: Math.round(dayData.main.temp_max),
        low: Math.round(dayData.main.temp_min),
        rain_chance: Math.round((dayData.pop || 0) * 100)
      });
    }

    const tips = [];
    const rainUpcoming = forecast.slice(0, 3).some(f => f.condition.toLowerCase().includes('rain'));
    if (rainUpcoming) {
      tips.push({ type: "danger", text: "🌧️ अगले कुछ दिनों में बारिश की संभावना है — अपनी फसल कटाई रोकें और सुरक्षित स्थान पर रखें।" });
      tips.push({ type: "warning", text: "💧 सिंचाई तुरंत बंद करें। खेतों में पानी जमा न होने दें।" });
    } else {
      if (current.main.temp > 35) {
        tips.push({ type: "warning", text: "🥵 अत्यधिक गर्मी — फसलों को सूखने से बचाने के लिए शाम को हल्की सिंचाई करें।" });
      }
      tips.push({ type: "safe", text: "🌤️ मौसम अनुकूल है — आज कीटनाशक छिड़काव (Pest spray) या यूरिया का प्रयोग करें।" });
      tips.push({ type: "safe", text: "🌾 कटी फसलों को सुखाने और मंडी ले जाने के लिए उपयुक्त समय है।" });
    }

    const now = new Date();
    const months = ["January", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December"];
    const formattedDate = `${now.getDate()} ${months[now.getMonth()]} ${now.getFullYear()}`;

    return res.json({
      weather: {
        location: cityName || location,
        date: formattedDate,
        temp: Math.round(current.main.temp),
        condition,
        humidity: current.main.humidity,
        wind_speed: Math.round(current.wind.speed * 3.6),
        pressure: current.main.pressure,
        visibility: Math.round((current.visibility || 10000) / 1000),
        forecast,
        farming_tips: tips
      }
    });

  } catch (error) {
    console.error('Weather error:', error.message);
    const location = (req.query.location || req.body.location || 'वाराणसी').trim();
    return res.json({ weather: getFallbackWeather(location) });
  }
});

router.get('/auto', async (req, res) => {
  const { lat, lon } = req.query;
  if (!process.env.OPENWEATHER_API_KEY || !lat || !lon) {
    return res.json({ city: "वाराणसी" });
  }

  try {
    const url = `http://api.openweathermap.org/geo/1.0/reverse?lat=${lat}&lon=${lon}&limit=1&appid=${process.env.OPENWEATHER_API_KEY}`;
    const response = await axios.get(url);
    if (response.data && response.data.length > 0) {
      return res.json({ city: response.data[0].name });
    }
  } catch (err) {
    console.error('Auto weather error:', err.message);
  }

  return res.json({ city: "वाराणसी" });
});

module.exports = router;
